public/app/plugins/datasource/prometheus/language_utils.ts

```diff
@@ -111,7 +112,55 @@ export function parseSelector(query: string, cursorOffset = 1): { labelKeys: any
 export function expandRecordingRules(query: string, mapping: { [name: string]: string }): string {
   const ruleNames = Object.keys(mapping);
   const rulesRegex = new RegExp(`(\\s|^)(${ruleNames.join('|')})(\\s|$|\\(|\\[|\\{)`, 'ig');
-  return query.replace(rulesRegex, (match, pre, name, post) => `${pre}${mapping[name]}${post}`);
+  const expandedQuery = query.replace(rulesRegex, (match, pre, name, post) => `${pre}${mapping[name]}${post}`);
+
+  // Split query into array, so if query uses operators, we can correctly add labels to each individual part.
+  const queryArray = expandedQuery.split(/(\+|\-|\*|\/|\%|\^)/);
+
+  // Regex that matches occurences of ){ or }{ or ]{ which is a sign of incorrecly added labels.
+  const invalidLabelsRegex = /(\)\{|\}\{|\]\{)/;
+  const correctlyExpandedQueryArray = queryArray.map(query => {
+    let expression = query;
+    if (expression.match(invalidLabelsRegex)) {
+      expression = addLabelsToExpression(expression, invalidLabelsRegex);
+    }
+    return expression;
+  });
+
+  return correctlyExpandedQueryArray.join('');
+}
+
+function addLabelsToExpression(expr: string, regex: RegExp) {
+  // Split query into 2 parts - before the invalidLabelsRegex match and after.
+  const indexOfRegexMatch = expr.match(regex).index;
+  const exprBeforeRegexMatch = expr.substr(0, indexOfRegexMatch + 1);
+  const exprAfterRegexMatch = expr.substr(indexOfRegexMatch + 1);
+
+  // Retrieve labels (as a string). Our exprAfterRegexMatch starts with "{" and we need to find closest matching bracket.
+  // We want to leave out the brackets and keep just string in labelsInString.
+  const stringOfLabels = exprAfterRegexMatch.substr(1, exprAfterRegexMatch.indexOf('}') - 1);
+
+  // Create array with labels
+  const arrayOfLabels = stringOfLabels.split(',');
+
+  // Transform strings into object with specified key, operator and value
+  // Uses the same regex and similar logic as function addLabelToSelector in add_label_to_query
+  const arrayOfLabelObjects = arrayOfLabels.map(string => {
+    const labelRegexp = /(\w+)\s*(=|!=|=~|!~)\s*("[^"]*")/g;
```

Comment:
There is already a similar regexp defined in line 47. Can we use that one, or fix it, or put a comment on why this needs a different one?